### PR TITLE
[core] Implement violation decorators

### LINKED ISSUE #3203: [core] Replace RuleViolationFactory implementations with ViolationDecorator
**Is your feature request related to a problem? Please describe.**
Currently [RuleViolation](https://github.com/pmd/pmd/blob/a8ad1ac4f6f873d0f2dd82e1ff2553c578ca6e02/pmd-core/src/main/java/net/sourceforge/pmd/RuleViolation.java#L89-L118) has a few methods which are somewhat language-specific:
* getPackageName
* getClassName
* getMethodName
* getVariableName

This is fine, but the way these properties are filled in is very indirect and weird: you need an implementation of RuleViolation which sets fields of its superclass in its constructor. You also need an implementation of RuleViolationFactory which calls the constructor of your subclass. So we multiply implementations of these interfaces even though ParametricRuleViolation, the default RuleViolation implementation already has all the necessary fields. Those implementations are useless except to fill those properties.

**Describe the solution you'd like**

* Expose a `Map<String, String>` as an attribute of RuleViolation, which contains key value pairs for these additional properties. Languages may add new properties at will.
* Add a parameter with this type to the constructor of ParametricRuleViolation.
* Add a new interface, `ViolationDecorator`, to pmd core. Violation decorators take a violation and add metadata to its attribute map. When creating a rule violation, we will just need to apply the registered decorator to add language-specific metadata, but the creating of the violation will not need any language-specific implementations. This way, RuleViolationFactory becomes redundant, and language-specific subclasses of RuleViolation too. So we reduce complexity when implementing a language. Here's a sketch of the interface:
```java
interface ViolationDecorator {
   RuleViolation decorate(RuleViolation baseViolation, Node node);
}
```
and an implementation:
```java
ViolationDecorator javaViolationDecorator = (baseViolation, node) -> {
    JavaNode node = (JavaNode) node;
    Map<String, String> metadata = new HashMap<>(baseViolation.getExtraAttributes()); // copy original map
    metadata.put(RuleViolation.VARIABLE_NAME, getVarName(javaNode));
    metadata.put(RuleViolation.METHOD_NAME, getMethodName(javaNode));
    // etc
    return new ParametricRuleViolation(baseViolation, metadata); // replace metadata, copy all other attributes of the base violation
};
```

**Describe alternatives you've considered** None

**Additional context** In #2807 I changed RuleViolationFactory a lot, with this idea in mind.

### LINKED ISSUE #3902: [core] Violation decorators
**Is your feature request related to a problem? Please describe.**

Currently violations have a couple of methods like getPackageName, getMethodName: https://github.com/pmd/pmd/blob/05d5bead37e04690d4f5169898801a6b1ec6dd1e/pmd-core/src/main/java/net/sourceforge/pmd/RuleViolation.java#L90-L119

These are mostly Java-specific and the way this data is filled-in is very awkward: we have a JavaRuleViolation subclass of RuleViolation that initializes the fields in its constructor. To create a JavaRuleViolation we need a  RuleViolationFactory instance which is mostly redundant.

The problems with this design are
* We need a subclass of ParametricRuleViolation even though the superclass already has all fields
* We need the interface RuleViolationFactory and language-specific implementations even though they're only used to call the correct constructor for the violation subclass
* The attributes cannot be extended easily by other languages than Java
* The attributes are meaningless in some languages, eg XML flavours

**Describe the solution you'd like**

* Refactor these "extra attributes" to be exposed through a method `Map<String, String> getAdditionalInfo()` in RuleViolation. Standard keys like PACKAGE_NAME and CLASS_NAME can be exposed as constants in the RuleViolation interface.
* Create a new interface `ViolationDecorator` whose job is to add extra attributes to the map. Use that in RuleContext when we create a new violation
* Remove the RuleViolation subclasses and finally remove RuleViolationFactory.

**Describe alternatives you've considered**

**Additional context**

- Part of PMD 7 #3898
- The new interface would be part of the `nspmd.reporting` package, like ViolationSuppressor
- The extra attributes are constrained to be string-valued because we need to serialize them to write them into the violation cache.

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

- Implement #3902 
- Remove RuleViolationFactory, and consolidate the whole logic to create and filter violations in RuleContext.

## Related issues

- Fixes #3902
- Fixes #3203

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 129 violations,<br />
introduces 28 new violations, 0 new errors and 0 new configuration errors,<br />
removes 23 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/a15f9d16af9241debafc57d9a65471f0/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 699661 new violations, 22 new errors and 0 new configuration errors,<br />
removes 827036 violations, 39 errors and 6 configuration errors.<br />
<a href="https://chunk.io/pmd/a15f9d16af9241debafc57d9a65471f0/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by oowekyala** [comment]
Regression: [AvoidUsingHardCodedIP](https://chunk.io/pmd/4d357d0d7a1a452ea13cae24cfa7cb3c/diff1/openjdk-11/index.html#rule-summary-AvoidUsingHardCodedIP) violation message is `Do not hard code the IP address ${variableName}`

**Review by adangel — APPROVED** [review]
Thanks!

I'll fix the issues and merge it afterwards.

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/CSVWriter.java [hunk 1]
```diff
@@ -16,13 +16,13 @@
  * @author Brian Remedios
  * @param <T>
  */
-public class CSVWriter<T extends Object> {
+class CSVWriter<T> {
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/CSVWriter.java:19]: Done: https://github.com/pmd/pmd/commit/8b7b0c7d5cdec78f1086a117e660a9d9c3e1fb8a

File: pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java [hunk 2]
```diff
@@ -30,26 +37,20 @@ public final class CachedRuleViolation implements RuleViolation {
...
 
     private CachedRuleViolation(final CachedRuleMapper mapper, final String description,
-            final String fileName, final String ruleClassName, final String ruleName,
-            final String ruleTargetLanguage, final int beginLine, final int beginColumn,
-            final int endLine, final int endColumn, final String packageName,
-            final String className, final String methodName, final String variableName) {
+                                final String fileName, final String ruleClassName, final String ruleName,
+                                final String ruleTargetLanguage, final int beginLine, final int beginColumn,
+                                final int endLine, final int endColumn,
+                                final Map<String, String> extraData) {
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java:46]: Done: https://github.com/pmd/pmd/commit/084a262071bed2c8f297d6a99b4ec7d3e57f07ae

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/YAHTMLRenderer.java [hunk 3]
```diff
@@ -49,7 +50,7 @@ public String defaultFileExtension() {
     }
 
     private void addViolation(RuleViolation violation) {
-        String packageName = violation.getPackageName();
+        String packageName = violation.getAdditionalInfo().getOrDefault(RuleViolation.PACKAGE_NAME, "");
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/YAHTMLRenderer.java:53]: Done: https://github.com/pmd/pmd/commit/be9f22ab5201744d26e6724a95a69178d045eab6

**Pull request merged**